### PR TITLE
Prometheus metric support label

### PR DESCRIPTION
**Pull request by hangc0276**
### Motivation
Current prometheus metric doesn't support configure scope label. If we configure multiple scope labels to a stats logger, it will append those labels into metric name, which leading metric name too long.

https://github.com/apache/bookkeeper/blob/74b00a101cb6ffd06dafccfe9d31c8c5be66efaf/bookkeeper-stats/src/main/java/org/apache/bookkeeper/stats/StatsLogger.java#L95-L104

### Modifications
1. Add label support for prometheus metrics

Current tests has been covered this change.
https://github.com/apache/pulsar/blob/4f1e39b6921ea401b8c27f17a041d06d85f8abf8/pulsar-broker/src/test/java/org/apache/pulsar/broker/stats/PrometheusMetricsTest.java#L969-L1008

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

- [x] `doc-not-needed`

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusTextFormatUtil.java [hunk 1]
```diff
@@ -167,4 +182,33 @@ public static void writeMetricsCollectedByPrometheusClient(Writer w, CollectorRe
...
+
+        boolean isFirst = true;
+        for (Map.Entry<String, String> e : labels.entrySet()) {
+            if (!isFirst) {
+                w.append(',');
+            }
+            isFirst = false;
+            w.append(e.getKey())
+                .append("=\"")
+                .append(e.getValue())
```
Comment (Shawyeok) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusTextFormatUtil.java:209]: In the future, the label (key or value) could be user input or configuration, is it safer if we do the escape job for special characters before write out? e.g.  `\` `"` `\n`, reference `io.prometheus.client.exporter.common.TextFormat#writeEscapedLabelValue`

Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/PrometheusTextFormatUtil.java:209]: Good idea +1

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/ScopeContext.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+package org.apache.pulsar.broker.stats.prometheus.metrics;
+
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Holder for a scope and a set of associated labels.
+ */
+public class ScopeContext {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/ScopeContext.java:27]: final ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/ScopeContext.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        this.scope = scope;
+        this.labels = labels;
+    }
+
+    public String getScope() {
+        return scope;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/metrics/ScopeContext.java:41]: We can use lombok here for equals and hashcode probably

**Review by eolivelli — CHANGES_REQUESTED** [review]
Good work.
There are a couple of comments.

Apart from those it looks like we don't have new tests that cover the new feature.
Existing tests arw good but we don't have specific test that verify the behaviour of labels, also with labels that contain special symbols

**Comment by BewareMyPower** [comment]
@hangc0276 Could you address the comments from @eolivelli ?

**Comment by Jason918** [comment]
@hangc0276 ping

**Comment by congbobo184** [comment]
@hangc0276   hi, I move this PR to release/2.9.5, if you have any questions, please ping me. thanks.

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**